R/toolCompareStatusLogs.R

```diff
@@ -0,0 +1,222 @@
+#' toolCompareStatusLogs
+#'
+#' Compares status logs and returns a textual rendering of the diff between the two.
+#'
+#' @param oldArchivePath Paths to the tgz-archive that serves as the old / new archive.
+#' If given, oldLogPath / newLogPath determines the name of the log file within the archive.
+#' If not given, oldLogPath / newLogPath need to be given.
+#' @param newArchivePath see above
+#' @param oldLogPath Paths to a log file, either within an archive,
+#' if an archive is given, or to a single file.
+#' @param newLogPath see above
+#' @param sections List of sections the output should include.
+#' Valid section names are: changedStatistics, changedCalls, addedCalls, removedCalls
+#' @returns A printable string describing the changes that occurred between
+#' the old and the new log.
+#' @author Patrick Rein
+#' @export
+toolCompareStatusLogs <- function(oldArchivePath = NULL, newArchivePath = NULL,
+                                  oldLogPath = file.path(".", "status.log"), newLogPath = file.path(".", "status.log"),
+                                  sections = c("changedStatistics", "changedCalls", "addedCalls", "removedCalls")) {
+  # Why is this a custom diff implementation instead of textual diff?
+  # -> Ordering in status.log can change in all kinds of ways and is not important.
+  #    Similar vein: statistics log entries have identity and should only be compared
+  #    based on their content.
+
+  if ((is.null(oldArchivePath) && is.null(newArchivePath) || oldArchivePath == newArchivePath)
+      && oldLogPath == newLogPath) {
+    # This is also a warning that none of the relevant parameters have been changed.
+    return("Nothing to do. Old and new paths are the same.")
+  }
+
+  tempDir <- withr::local_tempdir()
+  if (!is.null(oldArchivePath)) {
+    utils::untar(oldArchivePath,
+                 files = oldLogPath,
+                 exdir = file.path(tempDir, "old"))
+    oldStatusLog <- yaml::read_yaml(file.path(tempDir, "old", oldLogPath))
+  } else {
+    oldStatusLog <- yaml::read_yaml(oldLogPath)
+  }
+  if (!is.null(newArchivePath)) {
+    utils::untar(newArchivePath,
+                 files = newLogPath,
+                 exdir = file.path(tempDir, "new"))
+    newStatusLog <- yaml::read_yaml(file.path(tempDir, "new", newLogPath))
+  } else {
+    newStatusLog <- yaml::read_yaml(newLogPath)
+  }
+
+  diff <- .compareStatusLogsStatistics(oldStatusLog, newStatusLog)
+
+  return(.renderDiff(oldStatusLog, newStatusLog, diff, sections = sections))
+}
+
+.compareStatusLogsStatistics <- function(oldLog, newLog) {
+  # Assumed structure: list(string = call(list(entry)), entry = string|list
+
+  # Detect added/removed calls
+  removedCalls <- as.list(setdiff(names(oldLog), names(newLog)))
+  addedCalls <- as.list(setdiff(names(newLog), names(oldLog)))
+
+  # Detect changed calls
+  sameCalls <- as.list(intersect(names(newLog), names(oldLog)))
```

Comment:
@pascal-sauer Do you have a better solution for this pattern (L63-68)? I have used this several times now. It is basically a map that expands a vector of tokens into a named list with the tokens as keys and a values derived from the tokens as the values.